### PR TITLE
copr: Support collation aware like

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

###  What have you changed?

Based on #6562, this PR support collations in like built-in functions. See #6575

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

- Unit test

### CODE REVIEW

File: components/tidb_query/src/expr_util/collation/mod.rs [hunk 1]
```diff
@@ -55,7 +55,22 @@ impl From<DecodeOrWriteError> for crate::error::EvaluateError {
     }
 }
 
+pub trait Charset {
+    fn advance_one(data: &[u8]) -> usize;
```
Comment (sticnarf) [inline components/tidb_query/src/expr_util/collation/mod.rs]: The current implementation of `advance_one` looks like `next_char_width` to me... while an `advance_one` method should be something like `fn advance_one(data: &mut [u8]) -> usize` which should really advance the input slice and return the char codepoint.

I'm not sure what an optimal interface would be here.

**Comment by sticnarf** [comment]
And I think we'd better not rely on core internals if possible...What about this crate? https://crates.io/crates/encode_unicode

**Comment by TennyZhuang** [comment]
@sticnarf It is supported now.

File: components/tidb_query/src/expr_util/like.rs [hunk 2]
```diff
@@ -11,38 +11,48 @@
...
+                px += poff;
+                next_tx = tx
+                    + if let Some((_, toff)) = C::Charset::decode_one(&target[tx..]) {
+                        toff
+                    } else {
+                        1
+                    };
+                continue;
+            } else {
+                if code == escape && px + 1 < pattern.len() {
```
Comment (sticnarf) [inline components/tidb_query/src/expr_util/like.rs]: Do we still need `px + 1 < pattern.len()`?

Comment (TennyZhuang) [reply components/tidb_query/src/expr_util/like.rs]: fided

Comment (TennyZhuang) [reply components/tidb_query/src/expr_util/like.rs]: It's still necessary and should be fixed to `px + poff < pattern.len()`

Comment (sticnarf) [reply components/tidb_query/src/expr_util/like.rs]: AH... I thought wrong.

Comment (sticnarf) [reply components/tidb_query/src/expr_util/like.rs]: In MySQL:

```
mysql> select "C:\\" LIKE "C:\\";
+--------------------+
| "C:\\" LIKE "C:\\" |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.00 sec)
```
But the algorithm here gives `1`. The escape still has effect even if it's the last character.

Comment (TennyZhuang) [reply components/tidb_query/src/expr_util/like.rs]: I add a testcase 

```rust
(r#"C:\\"#, r#"C:\\"#, '\\', Collation::Binary, Some(0)),
```

and it passed.

Comment (sticnarf) [reply components/tidb_query/src/expr_util/like.rs]: It should be (r#"C:\"#, r#"C:\"#, '\\', Collation::Binary, Some(0)),

Comment (TennyZhuang) [reply components/tidb_query/src/expr_util/like.rs]: And I test it on Mysql 5.7

```SQL
SELECT VERSION();
SELECT "C:\\" LIKE "C:\\";
```

```
5.7.28-0ubuntu0.18.04.4
1

```

Comment (TennyZhuang) [reply components/tidb_query/src/expr_util/like.rs]: It seems that mysql will output 1 for this case?

Comment (TennyZhuang) [reply components/tidb_query/src/expr_util/like.rs]: Anyway, it seems that the change is bug-compatible with current implementation both on TiDB and TiKV. It seems that it's better to file a new bug issue for this case in TiDB?

Comment (TennyZhuang) [reply components/tidb_query/src/expr_util/like.rs]: I also tried:

```sql
SELECT VERSION();
select "x" LIKE "x" ESCAPE "x"
```

```
5.7.28-0ubuntu0.18.04.4
1
```

can't reproduce your result.

Comment (TennyZhuang) [reply components/tidb_query/src/expr_util/like.rs]: also 8.0

Comment (sticnarf) [reply components/tidb_query/src/expr_util/like.rs]: That's weird...

mysql> SELECT VERSION(); select 'x' LIKE 'x' ESCAPE "x";
+-----------+
| VERSION() |
+-----------+
| 5.7.28    |
+-----------+
1 row in set (0.00 sec)

+-------------------------+
| 'x' LIKE 'x' ESCAPE "x" |
+-------------------------+
|                       0 |
+-------------------------+
1 row in set (0.00 sec)

Comment (sticnarf) [reply components/tidb_query/src/expr_util/like.rs]: It's true on MySQL 8.0. Let's keep the behavior here.

**Pull request merged**